internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala

```
@@ -51,78 +52,115 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
-                  case pf: PlainFile => binaryDependency(pf.file, binaryClassName)
-                  case _             => ()
-                }
-              case None => ()
-            }
-          } else if (onSource.file != sourceFile) {
-            val onClassName = className(dep.to)
-            callback.classDependency(onClassName, fromClassName, context)
-          }
+      }
+      ()
+    }
+
+    // Define processor reusing `processDependency` definition
+    val memberRef = processDependency(DependencyByMemberRef) _
+    val inheritance = processDependency(DependencyByInheritance) _
+    val localInheritance = processDependency(LocalDependencyByInheritance) _
+
+    /*
+     * Handles dependency on given symbol by trying to figure out if represents a term
+     * that is coming from either source code (not necessarily compiled in this compilation
+     * run) or from class file and calls respective callback method.
+     */
+    def processDependency(context: DependencyContext)(dep: ClassDependency): Unit = {
+      val fromClassName = className(dep.from)
+
+      def binaryDependency(file: File, binaryClassName: String) =
+        callback.binaryDependency(file, binaryClassName, fromClassName, sourceFile, context)
+
+      import scala.tools.nsc.io.AbstractFile
+      def processExternalDependency(binaryClassName: String, at: AbstractFile) = {
+        at match {
+          case zipEntry: ZipArchive#Entry =>
+            // The dependency comes from a JAR
+            for {
+              zip <- zipEntry.underlyingSource
+              classFile <- Option(zip.file)
+            } binaryDependency(classFile, binaryClassName)
+          case pf: PlainFile =>
+            // The dependency comes from a class file
+            binaryDependency(pf.file, binaryClassName)
+          case _ =>
+          // TODO: If this happens, scala internals have changed. Log error.
```

This error is very unlikely to happen, but if it does it should be more than just `devWarning` or `warning`. That's why it's a todo.

---

Can we at least throw an exception for now?

---

I see that it is not a new logic so we can live without exception or error.